src/main/java/mods/railcraft/common/util/crafting/ShapelessFluidRecipe.java

```diff
@@ -0,0 +1,168 @@
+
+        for (int i = 0; i < var1.getSizeInventory(); i++) {
+            ItemStack slot = var1.getStackInSlot(i);
+
+            if (slot != null) {
+                boolean inRecipe = false;
+
+                for (Iterator<Object> iterator = required.iterator(); iterator.hasNext(); ) {
+                    boolean match = false;
+                    Object next = iterator.next();
+
+                    if (next instanceof ItemStack) {
+                        match = OreDictionary.itemMatches((ItemStack) next, slot, false);
+                    } else if (next instanceof List) {
+                        Iterator<ItemStack> itr = ((List<ItemStack>) next).iterator();
+                        while (itr.hasNext() && !match) {
+                            match = OreDictionary.itemMatches(itr.next(), slot, false);
+                        }
+                    } else if (next instanceof FluidStack) {
+                        if (slot.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, null)) {
+                            IFluidHandler handler = slot.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, null);
+                            FluidStack fluidStack = handler.drain((FluidStack) next, false);
+                            if (fluidStack == null || fluidStack.amount != ((FluidStack) next).amount) {
+                                continue;
+                            }
+                            match = true;
+                        } else if (slot.getItem() instanceof IFluidContainerItem) {
+                            // legacy
+                            IFluidContainerItem handler = (IFluidContainerItem) slot.getItem();
+                            FluidStack fluidStack = handler.getFluid(slot);
+                            if (fluidStack == null || fluidStack.getFluid() != ((FluidStack) next).getFluid() || fluidStack.amount < ((FluidStack) next).amount) {
+                                continue;
+                            }
+                            match = true;
+                        } else {
+                            continue;
+                        }
+                    }
+
+                    if (match) {
+                        inRecipe = true;
+                        iterator.remove();
+                        break;
+                    }
+                }
+
+                if (!inRecipe) {
+                    return false;
+                }
+            }
+        }
+
+        return required.isEmpty();
+    }
+
+    /**
+     * Returns the input for this recipe, any mod accessing this value should never
+     * manipulate the values in this array as it will effect the recipe itself.
+     *
+     * @return The recipes input vales.
+     */
+    public List<Object> getInput() {
+        return this.input;
+    }
+
+    @Override
+    public ItemStack[] getRemainingItems(InventoryCrafting inv) {
+        return ForgeHooks.defaultRecipeGetRemainingItems(inv);
```

Comment:
If this is copied directly from my PR, then this part is incomplete. It needs to find and drain the correct amount from the containers.

Reply:
Will adjust that. Or can I keep a list after the `getResult` call, presuming that this is always called after getting the result?

Reply:
Done. I checked in `matches` and used that drain results.